nntrainer/src/concat_layer.cpp

```diff
@@ -0,0 +1,124 @@
+ *
+ */
+
+#include <concat_layer.h>
+#include <cstring>
+#include <layer_internal.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <parse_util.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+int ConcatLayer::initialize() {
+  int status = ML_ERROR_NONE;
+  unsigned int channel = 0;
+  if (num_inputs == 0) {
+    ml_loge("Error: number of inputs are not initialized");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+    channel += input_dim[idx].channel();
+  }
+
+  output_dim[0] = input_dim[0];
+  output_dim[0].channel(channel);
+
+  return status;
+}
+
+sharedConstTensors ConcatLayer::forwarding(sharedConstTensors in) {
+
+  TensorDim d = in[0]->getDim();
+  unsigned int concat_channel = d.channel();
+
+  for (unsigned int idx = 1; idx < num_inputs; ++idx) {
+    TensorDim dim = in[idx]->getDim();
+
+    for (unsigned int i = 2; i < d.rank(); ++i) {
+      if (d[i] != dim[i])
+        throw std::runtime_error("Error: concat layer requires same "
+                                 "shape from  all input layers");
+    }
+    concat_channel += dim.channel();
+  }
+
+  if (concat_channel != output_dim[0].channel()) {
+    throw std::runtime_error(
+      "Error: Sum of channel of input layers is not same with output channel");
+  }
+
+  hidden = Tensor(output_dim[0]);
+  hidden.setZero();
+  unsigned int f_size = d.width() * d.height() * (concat_channel);
+
+  for (unsigned int b = 0; b < d.batch(); ++b) {
+    unsigned int position = 0;
+    for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+      TensorDim in_dim = in[idx]->getDim();
+      memcpy(hidden.getAddress(b * f_size + position),
+             in[idx]->getAddress(b * in_dim.getFeatureLen()),
+             in_dim.getFeatureLen() * sizeof(float));
+      position += in_dim.getFeatureLen();
+    }
+  }
+
+  return {MAKE_SHARED_TENSOR(hidden)};
+}
+
+sharedConstTensors ConcatLayer::backwarding(sharedConstTensors derivative,
+                                            int iteration) {
+  sharedConstTensors ret;
+  TensorDim d = derivative[0]->getDim();
+
+  unsigned int position = 0;
+  for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+    TensorDim in_dim = input_dim[idx];
+    sharedTensor t = std::shared_ptr<Tensor>(new Tensor(in_dim),
+                                             std::default_delete<Tensor[]>());
+
+    for (unsigned int b = 0; b < in_dim.batch(); ++b) {
+      memcpy(t->getAddress(b * in_dim.getFeatureLen()),
+             derivative[0]->getAddress(b * d.getFeatureLen() + position),
+             in_dim.getFeatureLen() * sizeof(float));
+    }
+    position += in_dim.getFeatureLen();
+
+    ret.push_back(t);
+  }
+
+  return ret;
+}
+
+void ConcatLayer::setProperty(const PropertyType type,
+                              const std::string &value) {
+  int status = ML_ERROR_NONE;
+
+  switch (type) {
+  case PropertyType::num_inputs: {
```

Comment:
IMO, it is more natural to set num_inputs from parsing input_layers in and disallow setting for `num_inputs`

```ini
[concat1]
Type = concat
input_layers ="input_layer1, input_layer2"
```

Reply:
This num_inputs may not required according to following implementation. I will change this as it implemented.